File: src/struct/Map.java
package struct;

import annotation.Intention;
import annotation.Required;
import annotation.Volatile;
import lang.Class;
import nio.ForeignMemory;
import nio.MemoryRegistry;
import oop.TypeRegister;
import util.Hash;

import java.lang.foreign.Arena;

/**
 * Off-heap key-value hash map implementation.
 */
@Volatile
@Intention("Zero-GC off-heap open-addressing hash map with key class inspection, 64-bit hashing, thread-safe memory mutations, and global MemoryRegistry integration.")
public final class Map {

    @Required
    public static final int CLASS_ID = TypeRegister.ID_MAP;

    public static final int TYPE_MAP = TypeRegister.FORM_POINTER | CLASS_ID; // 0xCC000013

    private static final int DEFAULT_CAPACITY = 16;
    private static final float LOAD_FACTOR = 0.75f;
    private static final long HEADER_SIZE = 24L; // 8B metadata header + 16B slot layout
    private static final long SLOT_SIZE = 32L;   // 8B key + 8B val + 8B hash + 8B state (0=empty, 1=occupied, 2=deleted)

    private static final int STATE_EMPTY = 0;
    private static final int STATE_OCCUPIED = 1;
    private static final int STATE_DELETED = 2;

    private static Arena poolArena;
    private static volatile boolean active;

    static {
        poolArena = Arena.ofShared();
        active = true;
        MemoryRegistry.register(Map::freeAll);
    }

    private Map() {}

    private static void checkActive() {
        if (!active) throw new IllegalStateException("Map subsystem is not active!");
    }

    private static void checkValid(long mapPtr) {
        if (mapPtr == 0L) throw new NullPointerException("Accessing NULL off-heap map pointer!");
    }

    // free all subsystem resources
    public static void freeAll() {
        if (active) {
            active = false;
            if (poolArena != null && poolArena.scope().isAlive()) {
                poolArena.close();
            }
        }
    }

    // create empty off-heap map for keyClassId and valClassId
    public static long instant(int keyClassId, int valClassId) {
        return instant(keyClassId, valClassId, DEFAULT_CAPACITY);
    }

    // create empty off-heap map with initial capacity
    public static long instant(int keyClassId, int valClassId, int initialCapacity) {
        checkActive();
        int cap = (initialCapacity <= 0) ? DEFAULT_CAPACITY : highestOneBit(initialCapacity);
        if (cap < 4) cap = 4;

        long headerBlock = ForeignMemory.allocateNative(HEADER_SIZE);
        long userPtr = headerBlock + 8L;

        ForeignMemory.putInt(headerBlock, TYPE_MAP);
        ForeignMemory.putInt(headerBlock + 4L, 0); // activeCount

        ForeignMemory.putInt(userPtr, keyClassId);
        ForeignMemory.putInt(userPtr + 4L, valClassId);
        ForeignMemory.putInt(userPtr + 8L, cap);
        ForeignMemory.putInt(userPtr + 12L, 0); // padding

        long bufferBytes = (long) cap * SLOT_SIZE;
        long dataBuffer = ForeignMemory.allocateNative(bufferBytes);
        ForeignMemory.setMemory(dataBuffer, bufferBytes, (byte) 0);
        ForeignMemory.putLong(userPtr + 16L, dataBuffer);

        return userPtr;
    }

    private static boolean isReferenceClass(int classId) {
        return classId == TypeRegister.ID_STRING || classId >= TypeRegister.ID_LIST;
    }

    // compute 64-bit hash for key based on key class inspection
    private static long computeHash(int keyClassId, long key) {
        if (key == 0L) return 0L;
        if (isReferenceClass(keyClassId) && key >= 4096L) {
            int inspectedClass = Class.getClass(key);
            if (inspectedClass != 0) {
                int len = Class.getLength(key);
                if (len > 0) return Hash.fnv1a64(key, len);
            }
        }
        return Hash.murmur3Mix64(key);
    }

    // compare two keys for equality based on class inspection
    private static boolean keysEqual(int keyClassId, long k1, long k2) {
        if (k1 == k2) return true;
        if (k1 == 0L || k2 == 0L) return false;
        if (isReferenceClass(keyClassId) && k1 >= 4096L && k2 >= 4096L) {
            int c1 = Class.getClass(k1);
            int c2 = Class.getClass(k2);
            if (c1 != 0 && c1 == c2) {
                int len1 = Class.getLength(k1);
                int len2 = Class.getLength(k2);
                if (len1 != len2) return false;
                for (int i = 0; i < len1; i++) {
                    if (ForeignMemory.getByte(k1 + i) != ForeignMemory.getByte(k2 + i)) {
                        return false;
                    }
                }
                return true;
            }
        }
        return false;
    }


    // put key-value entry into map
    public static synchronized void put(long mapPtr, long key, long value) {
        checkActive();
        checkValid(mapPtr);

        int count = size(mapPtr);
        int cap = capacity(mapPtr);
        if (count >= (int) (cap * LOAD_FACTOR)) {
            rehash(mapPtr, cap * 2);
            cap = capacity(mapPtr);
        }

        int keyClassId = keyClassId(mapPtr);
        long hash = computeHash(keyClassId, key);
        long buffer = dataBuffer(mapPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);
        int firstDeleted = -1;

        while (true) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 24L);

            if (st == STATE_EMPTY) {
                int targetIdx = (firstDeleted != -1) ? firstDeleted : idx;
                long targetSlot = buffer + ((long) targetIdx * SLOT_SIZE);
                ForeignMemory.putLong(targetSlot, key);
                ForeignMemory.putLong(targetSlot + 8L, value);
                ForeignMemory.putLong(targetSlot + 16L, hash);
                ForeignMemory.putLong(targetSlot + 24L, STATE_OCCUPIED);
                ForeignMemory.putInt(mapPtr - 4L, count + 1);
                return;
            } else if (st == STATE_DELETED) {
                if (firstDeleted == -1) firstDeleted = idx;
            } else if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 16L);
                long slotKey = ForeignMemory.getLong(slot);
                if (slotHash == hash && keysEqual(keyClassId, slotKey, key)) {
                    ForeignMemory.putLong(slot + 8L, value); // update value
                    return;
                }
            }
            idx = (idx + 1) & mask;
        }
    }

    // get value for key, returns 0L if not found
    public static synchronized long get(long mapPtr, long key) {
        checkValid(mapPtr);
        int cap = capacity(mapPtr);
        if (cap == 0) return 0L;

        int keyClassId = keyClassId(mapPtr);
        long hash = computeHash(keyClassId, key);
        long buffer = dataBuffer(mapPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);

        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 24L);

            if (st == STATE_EMPTY) return 0L;
            if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 16L);
                long slotKey = ForeignMemory.getLong(slot);
                if (slotHash == hash && keysEqual(keyClassId, slotKey, key)) {
                    return ForeignMemory.getLong(slot + 8L);
                }
            }
            idx = (idx + 1) & mask;
        }
        return 0L;
    }

    // check if map contains key
    public static synchronized boolean containsKey(long mapPtr, long key) {
        checkValid(mapPtr);
        int cap = capacity(mapPtr);
        if (cap == 0) return false;

        int keyClassId = keyClassId(mapPtr);
        long hash = computeHash(keyClassId, key);
        long buffer = dataBuffer(mapPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);

        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 24L);

            if (st == STATE_EMPTY) return false;
            if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 16L);
                long slotKey = ForeignMemory.getLong(slot);
                if (slotHash == hash && keysEqual(keyClassId, slotKey, key)) {
                    return true;
                }
            }
            idx = (idx + 1) & mask;
        }
        return false;
    }

    // remove key entry from map and return old value
    public static synchronized long remove(long mapPtr, long key) {
        checkValid(mapPtr);
        int cap = capacity(mapPtr);
        if (cap == 0) return 0L;

        int keyClassId = keyClassId(mapPtr);
        long hash = computeHash(keyClassId, key);
        long buffer = dataBuffer(mapPtr);
        int mask = cap - 1;
        int idx = (int) (hash & mask);

        for (int i = 0; i < cap; i++) {
            long slot = buffer + ((long) idx * SLOT_SIZE);
            long st = ForeignMemory.getLong(slot + 24L);

            if (st == STATE_EMPTY) return 0L;
            if (st == STATE_OCCUPIED) {
                long slotHash = ForeignMemory.getLong(slot + 16L);
                long slotKey = ForeignMemory.getLong(slot);
                if (slotHash == hash && keysEqual(keyClassId, slotKey, key)) {
                    long oldVal = ForeignMemory.getLong(slot + 8L);
                    ForeignMemory.putLong(slot + 24L, STATE_DELETED);
                    int count = size(mapPtr);
                    ForeignMemory.putInt(mapPtr - 4L, count - 1);
                    return oldVal;
                }
            }
            idx = (idx + 1) & mask;
        }
        return 0L;
    }

    private static void rehash(long mapPtr, int newCap) {
        int oldCap = capacity(mapPtr);
        long oldBuffer = dataBuffer(mapPtr);
        long newBufferBytes = (long) newCap * SLOT_SIZE;
        long newBuffer = ForeignMemory.allocateNative(newBufferBytes);
        ForeignMemory.setMemory(newBuffer, newBufferBytes, (byte) 0);

        int mask = newCap - 1;
        for (int i = 0; i < oldCap; i++) {
            long slot = oldBuffer + ((long) i * SLOT_SIZE);
            if (ForeignMemory.getLong(slot + 24L) == STATE_OCCUPIED) {
                long key = ForeignMemory.getLong(slot);
                long val = ForeignMemory.getLong(slot + 8L);
                long hash = ForeignMemory.getLong(slot + 16L);

                int idx = (int) (hash & mask);
                while (ForeignMemory.getLong(newBuffer + ((long) idx * SLOT_SIZE) + 24L) == STATE_OCCUPIED) {
                    idx = (idx + 1) & mask;
                }
                long targetSlot = newBuffer + ((long) idx * SLOT_SIZE);
                ForeignMemory.putLong(targetSlot, key);
                ForeignMemory.putLong(targetSlot + 8L, val);
                ForeignMemory.putLong(targetSlot + 16L, hash);
                ForeignMemory.putLong(targetSlot + 24L, STATE_OCCUPIED);
            }
        }

        ForeignMemory.freeNative(oldBuffer);
        ForeignMemory.putLong(mapPtr + 16L, newBuffer);
        ForeignMemory.putInt(mapPtr + 8L, newCap);
    }

    private static int highestOneBit(int i) {
        return Integer.highestOneBit(i - 1) << 1;
    }

    // check if map is empty
    public static boolean isEmpty(long mapPtr) {
        return Collection.isEmpty(mapPtr);
    }

    // get element count
    public static int size(long mapPtr) {
        return Collection.size(mapPtr);
    }

    // free map data buffer and header back to native RAM
    public static synchronized void free(long mapPtr) {
        checkActive();
        if (mapPtr == 0L) return;

        long headerBlock = mapPtr - 8L;
        int type = ForeignMemory.getInt(headerBlock);
        if (type == 0 || !TypeRegister.isPointer(type)) {
            throw new IllegalStateException("Double free or corrupt map pointer: 0x" + Long.toHexString(mapPtr).toUpperCase());
        }

        long buffer = dataBuffer(mapPtr);
        if (buffer != 0L) {
            ForeignMemory.freeNative(buffer);
        }

        ForeignMemory.putInt(headerBlock, 0);
        ForeignMemory.putInt(headerBlock + 4L, -1);
        ForeignMemory.freeNative(headerBlock);
    }

    public static int keyClassId(long mapPtr) {
        return Collection.keyClassId(mapPtr);
    }

    public static int valClassId(long mapPtr) {
        return Collection.valClassId(mapPtr);
    }

    public static int capacity(long mapPtr) {
        return Collection.capacity(mapPtr);
    }

    public static long dataBuffer(long mapPtr) {
        return Collection.dataBuffer(mapPtr);
    }

    public static int classId() {
        return CLASS_ID;
    }
}

File: src/struct/map.c
#include "struct/map.h"

#include <string.h>

#include "nio/mem.h"
#include "oop/type.h"
#include "util/hash.h"
#include "annotation/overview.h"

;;OVERVIEW
/**
 * ============================================================================
 * CLASS: Map (struct/map.c)
 * LEVEL: L2 — Behavior (container behavior API)
 * ============================================================================
 * the Map class, ported from struct/Map.java.
 *
 * STRUCT FIELDS (Mirroring struct/map.h):
 * ----------------------------------------------------------------------------
 *   Map {
 *     Collection collection; // instance state
 *   }
 *
 * FUNCTION REGISTRY:
 * ----------------------------------------------------------------------------
 * Constructors:
 *   - Map_3(key_class, val_class, capacity)
 *   - Map_2(key_class, val_class)
 *
 * Core Functions:
 *   - Map_free(map)
 *   - Map_put(map, key, value)
 *   - Map_containsKey(map, key)
 *   - Map_remove(map, key)
 *   - Map_keys(map)
 *   - Map_size(map)
 *   - Map_capacity(map)
 *   - Map_keyClassId(map)
 *   - Map_valClassId(map)
 *   - Map_dataBuffer(map)
 *
 * Getters:
 *   - Map_get(map, key)
 *   - Map_isEmpty(map)
 * ============================================================================
 */


// map.c — Map port (Legacy: struct/Map.java).

static const size_t SLOT_SIZE = 32;
static const size_t DEFAULT_CAPACITY = 16;
static const size_t LOAD_DIVISOR = 4; // 0.75 load factor = capacity - capacity/4
static const uint64_t STATE_EMPTY = 0;
static const uint64_t STATE_OCCUPIED = 1;
static const uint64_t STATE_DELETED = 2;

static Collection *asCollection(Map *map) {
    return (Collection*) map;
}

static int isReferenceClass(uint32_t classId) {
    return classId == ID_STRING || classId >= ID_LIST;
}

static uint64_t computeHash(uint32_t keyClass, uint64_t key) {
    if (key == 0) return 0;
    if (isReferenceClass(keyClass) && key >= 4096u) {
        uint64_t inspected = Memory_type((void*) (uintptr_t)key);
        if (inspected != 0) {
            size_t len = Memory_length((void*) (uintptr_t)key);
            if (len > 0)
                return Hash_fnv1a64((const uint8_t*) (uintptr_t)key, len);
        }
    }
    return Hash_murmur3Mix64(key);
}

static int keysEqual(uint32_t keyClass, uint64_t k1, uint64_t k2) {
    if (k1 == k2) return 1;
    if (k1 == 0 || k2 == 0) return 0;
    if (isReferenceClass(keyClass) && k1 >= 4096u && k2 >= 4096u) {
        void *p1 = (void*) (uintptr_t)k1;
        void *p2 = (void*) (uintptr_t)k2;
        uint64_t t1 = Memory_type(p1);
        uint64_t t2 = Memory_type(p2);
        if (t1 != 0 && t1 == t2) {
            size_t len1 = Memory_length(p1);
            size_t len2 = Memory_length(p2);
            if (len1 != len2) return 0;
            return memcmp(p1, p2, len1) == 0;
        }
    }
    return 0;
}

static size_t highestOneBit(size_t n) {
    if (n <= 1) return 1;
    size_t v = n - 1;
    v |= v >> 1;
    v |= v >> 2;
    v |= v >> 4;
    v |= v >> 8;
    v |= v >> 16;
    v |= v >> 32;
    return v + 1;
}

static uint8_t *slotAt(Collection *c, size_t index) {
    return (*c).data + index * SLOT_SIZE;
}

static uint64_t slotState(const uint8_t *slot) {
    return *(uint64_t*) (slot + 24);
}

static void rehash(Collection *c, size_t newCap) {
    size_t oldCap = (*c).capacity;
    uint8_t *oldData = (*c).data;
    size_t bytes = newCap * SLOT_SIZE;
    uint64_t bufType = Type_make(PROJ_VEXSPOKE, FORM_ARRAY, ID_MAP);
    uint8_t *newData = (uint8_t*) Memory_alloc(bufType, bytes);
    if (!newData) return;
    memset(newData, 0, bytes);

    size_t mask = newCap - 1;
    for (size_t i = 0; i < oldCap; i++) {
        uint8_t *slot = slotAt(c, i);
        if (slotState(slot) == STATE_OCCUPIED) {
            uint64_t key = *(uint64_t*) slot;
            uint64_t val = *(uint64_t*) (slot + 8);
            uint64_t hash = *(uint64_t*) (slot + 16);

            size_t idx = (size_t)(hash & mask);
            while (slotState(newData + idx * SLOT_SIZE) == STATE_OCCUPIED)
                idx = (idx + 1) & mask;
            uint8_t *target = newData + idx * SLOT_SIZE;
            *(uint64_t*) target = key;
            *(uint64_t*) (target + 8) = val;
            *(uint64_t*) (target + 16) = hash;
            *(uint64_t*) (target + 24) = STATE_OCCUPIED;
        }
    }

    Memory_free(oldData);
    (*c).data = newData;
    (*c).capacity = (uint32_t)newCap;
}

Map *Map_3(uint32_t keyClass, uint32_t valClass, size_t capacity) {
    size_t cap = capacity == 0 ? DEFAULT_CAPACITY : highestOneBit(capacity);
    if (cap < 4) cap = 4;

    Map *map = (Map*) Memory_alloc(TYPE_MAP, sizeof(Map));
    if (!map) return nullptr;

    Collection *c = asCollection(map);
    (*c).typeId = TYPE_MAP;
    (*c).activeCount = 0;
    (*c).elementClass = keyClass;
    (*c).stride = valClass;
    (*c).capacity = (uint32_t)cap;
    (*c).head = 0;

    size_t bytes = cap * SLOT_SIZE;
    uint64_t bufType = Type_make(PROJ_VEXSPOKE, FORM_ARRAY, ID_MAP);
    (*c).data = (uint8_t*) Memory_alloc(bufType, bytes);
    if (!(*c).data) {
        Memory_free(map);
        return nullptr;
    }
    memset((*c).data, 0, bytes);
    return map;
}

void Map_free(Map *map) {
    if (!map) return;
    Collection *c = asCollection(map);
    if ((*c).data)
        Memory_free((*c).data);
    Memory_free(map);
}

void Map_put(Map *map, uint64_t key, uint64_t value) {
    if (!map) return;
    Collection *c = asCollection(map);

    size_t load = (*c).capacity - (*c).capacity / LOAD_DIVISOR;
    if ((*c).activeCount >= load)
        rehash(c, (*c).capacity * 2);

    uint64_t hash = computeHash((*c).elementClass, key);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);
    size_t firstDeleted = SIZE_MAX;

    while (true) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);

        if (st == STATE_EMPTY) {
            size_t target = firstDeleted != SIZE_MAX ? firstDeleted : idx;
            uint8_t *tslot = slotAt(c, target);
            *(uint64_t*) tslot = key;
            *(uint64_t*) (tslot + 8) = value;
            *(uint64_t*) (tslot + 16) = hash;
            *(uint64_t*) (tslot + 24) = STATE_OCCUPIED;
            (*c).activeCount++;
            return;
        } else if (st == STATE_DELETED) {
            if (firstDeleted == SIZE_MAX)
                firstDeleted = idx;
        } else if (st == STATE_OCCUPIED) {
            if (*(uint64_t*) (slot + 16) == hash
                && keysEqual((*c).elementClass, *(uint64_t*) slot, key)) {
                *(uint64_t*) (slot + 8) = value;
                return;
            }
        }
        idx = (idx + 1) & mask;
    }
}

uint64_t Map_get(Map *map, uint64_t key) {
    if (!map) return 0;
    Collection *c = asCollection(map);
    if ((*c).capacity == 0) return 0;

    uint64_t hash = computeHash((*c).elementClass, key);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);
        if (st == STATE_EMPTY) return 0;
        if (st == STATE_OCCUPIED
            && *(uint64_t*) (slot + 16) == hash
            && keysEqual((*c).elementClass, *(uint64_t*) slot, key))
            return *(uint64_t*) (slot + 8);
        idx = (idx + 1) & mask;
    }
    return 0;
}

bool Map_containsKey(Map *map, uint64_t key) {
    if (!map) return false;
    Collection *c = asCollection(map);
    if ((*c).capacity == 0) return false;

    uint64_t hash = computeHash((*c).elementClass, key);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);
        if (st == STATE_EMPTY) return false;
        if (st == STATE_OCCUPIED
            && *(uint64_t*) (slot + 16) == hash
            && keysEqual((*c).elementClass, *(uint64_t*) slot, key))
            return true;
        idx = (idx + 1) & mask;
    }
    return false;
}

uint64_t Map_remove(Map *map, uint64_t key) {
    if (!map) return 0;
    Collection *c = asCollection(map);
    if ((*c).capacity == 0) return 0;

    uint64_t hash = computeHash((*c).elementClass, key);
    size_t mask = (*c).capacity - 1;
    size_t idx = (size_t)(hash & mask);

    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, idx);
        uint64_t st = slotState(slot);
        if (st == STATE_EMPTY) return 0;
        if (st == STATE_OCCUPIED
            && *(uint64_t*) (slot + 16) == hash
            && keysEqual((*c).elementClass, *(uint64_t*) slot, key)) {
            uint64_t oldVal = *(uint64_t*) (slot + 8);
            *(uint64_t*) (slot + 24) = STATE_DELETED;
            (*c).activeCount--;
            return oldVal;
        }
        idx = (idx + 1) & mask;
    }
    return 0;
}

Array *Map_keys(Map *map) {
    if (!map) return nullptr;
    Collection *c = asCollection(map);
    Array *keys = Array((*c).elementClass, (*c).activeCount);
    if (!keys) return nullptr;

    size_t out = 0;
    for (size_t i = 0; i < (*c).capacity; i++) {
        uint8_t *slot = slotAt(c, i);
        if (slotState(slot) == STATE_OCCUPIED)
            Array_set(keys, out++, *(uint64_t*) slot);
    }
    return keys;
}

bool Map_isEmpty(Map *map) {
    return Collection_isEmpty(asCollection(map));
}

size_t Map_size(Map *map) {
    return Collection_size(asCollection(map));
}

size_t Map_capacity(Map *map) {
    return Collection_capacity(asCollection(map));
}

uint32_t Map_keyClassId(Map *map) {
    return Collection_keyClassId(asCollection(map));
}

uint32_t Map_valClassId(Map *map) {
    return Collection_valClassId(asCollection(map));
}

uint8_t *Map_dataBuffer(Map *map) {
    return Collection_dataBuffer(asCollection(map));
}
Map *Map_2(uint32_t key_class, uint32_t val_class) {
    return Map_3(key_class, val_class, 16);
}
